govuk_chat_evaluation/question_router/evaluate.py

```diff
@@ -0,0 +1,182 @@
+        )
+
+    @cached_property
+    def _expected_actual_lists(self) -> tuple[list[str], list[str]]:
+        pairs_list = [
+            (eval.expected_outcome, eval.actual_outcome)
+            for eval in self.evaluation_results
+        ]
+
+        expected, actual = zip(*pairs_list)
+        return list(expected), list(actual)
+
+    def accuracy(self) -> float:
+        return accuracy_score(
+            *self._expected_actual_lists,  # type: ignore
+        )
+
+    def precision(self) -> float:
+        return precision_score(
+            *self._expected_actual_lists,
+            average="weighted",
+            zero_division=np.nan,  # type: ignore
+        )
+
+    def recall(self) -> float:
+        return recall_score(
+            *self._expected_actual_lists,
+            average="weighted",
+            zero_division=np.nan,  # type: ignore
+        )
+
+    def f1_score(self) -> float:
+        return f1_score(
+            *self._expected_actual_lists,
+            average="weighted",
+            zero_division=np.nan,  # type: ignore
+        )
+
+    def f2_score(self) -> float:
+        return fbeta_score(
+            *self._expected_actual_lists,
+            beta=2,
+            average="weighted",
+            zero_division=np.nan,  # type: ignore
+        )
+
+    def confusion_matrix_data(self) -> list[list[int]]:
+        return confusion_matrix(
+            *self._expected_actual_lists,
+            labels=sorted(list(set(self.classification_labels))),  # type: ignore
+        )
+
+    def miscategorised_cases(self) -> list[dict[str, Any]]:
+        return [
+            {
+                "question": result.question,
+                "predicted_classification": result.expected_outcome,
+                "actual_classification": result.actual_outcome,
+                "confidence_score": result.confidence_score,
+            }
+            for result in self.evaluation_results
+            if result.expected_outcome != result.actual_outcome
+        ]
+
+    def to_dict(self) -> dict[str, Any]:
+        return {
+            "Evaluated": len(self.evaluation_results),
+            "Accuracy": self.accuracy(),
+            "Precision": self.precision(),
+            "Recall": self.recall(),
+            "F1 Score": self.f1_score(),
+            "F2 Score": self.f2_score(),
+            "Miscategorised Cases": len(self.miscategorised_cases()),
+        }
+
+    def for_csv(self) -> list[dict[str, Any]]:
+        return [{"property": k, "value": v} for k, v in self.to_dict().items()]
+
+
+def generate_and_output_confusion_matrix(
+    output_dir: Path,
+    confusion_matrix_data: list[list[int]],
+    confusion_matrix_labels: list[str],
+):
+    # Takes confusion matrix data (a 2D list) calculated by sklearn
+    # and a list of labels (strings representing the question routing labels)
+    # and outputs an confusion matrix PNG image to the output directory
```

Comment:
Content looks good, but I think it'd be more pythonic to be in the triple quotes string